src/formats/CIF.cpp

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) Guillaume Fraux and contributors -- BSD license
+
+#include "chemfiles/config.h"
+#ifndef CHFL_DISABLE_GEMMI
+
+#include <cassert>
+#include <cstdint>
+#include <map>
+#include <array>
+#include <vector>
+#include <string>
+#include <memory>
+#include <utility>
+#include <algorithm>
+
+#include "chemfiles/File.hpp"
+#include "chemfiles/Format.hpp"
+#include "chemfiles/Atom.hpp"
+#include "chemfiles/Frame.hpp"
+#include "chemfiles/Property.hpp"
+#include "chemfiles/Residue.hpp"
+#include "chemfiles/Topology.hpp"
+#include "chemfiles/UnitCell.hpp"
+
+#include "chemfiles/types.hpp"
+#include "chemfiles/utils.hpp"
+#include "chemfiles/parse.hpp"
+#include "chemfiles/error_fmt.hpp"
+#include "chemfiles/string_view.hpp"
+#include "chemfiles/external/optional.hpp"
+
+#include "chemfiles/formats/CIF.hpp"
+
+using namespace chemfiles;
+
+template<> FormatInfo chemfiles::format_information<CIFFormat>() {
+    return FormatInfo("CIF").with_extension(".cif").description(
+        "CIF (Crystallographic Information Framework)"
+    );
+}
+
+void CIFFormat::init_() {
+    gemmi::cif::Document *doc;
+
+    if (file_.mode() == File::WRITE) {
+        return;
+    }
+
+    if (file_.mode() == File::APPEND) {
+        throw file_error("cannot open CIF files in append ('a') mode");
+    }
+
+    // Parse the CIF file
+    auto content = file_.readall();
+    try {
+	doc = new gemmi::cif::Document(gemmi::cif::read_string(content));
+    } catch (std::exception& e) {
+	throw format_error("cannot parse CIF file");
+    }
+
+    // Interpret the content of each block, to remove blocks that contain
+    // no atom: those can occur, but are not useful to us (they contain comments
+    // and associated experimental data).
+    for (const auto &block: doc->blocks) {
+      try {
+	gemmi::SmallStructure s = gemmi::make_small_structure_from_block(block);
+
+	if (s.get_all_unit_cell_sites().size() > 0) {
+	  structures_.push_back(s);
+	}
+      } catch (std::exception& e) {
+	/* Simply skip to the next block. */
+	throw format_error("cannot interpret CIF block");
+      }
+    }
+
+    // TODO: if the CIF file has zero blocks with atoms, maybe we should throw an error here
+    // if (structures_.size() == 0) ...
+}
+
+size_t CIFFormat::nsteps() {
+    return structures_.size();
+}
+
+void CIFFormat::read_step(const size_t step, Frame& frame) {
+    assert(step < structures_.size());
+
+    auto structure = structures_[step];
+    auto sites = structure.get_all_unit_cell_sites();
+
+    UnitCell cell = UnitCell(structure.cell.a, structure.cell.b, structure.cell.c,
+	structure.cell.alpha, structure.cell.beta, structure.cell.gamma);
+    frame.set_cell(cell);
+
+    frame.set("name", structure.name);
+
+    for (const auto& site: sites) {
+        Atom atom(site.label, site.element.name());
+        // TODO: should we use those somehow?
+	// site.element.atomic_number()
+	// site.element.weight()
+	// site.occ is the occupancy
```

Comment:
weight can be used as the atom mass `atom.set_mass(site.weight())`, with the caveat that chemfiles expects masses in atomic mass units. I don't know what's the standard in CIF files

for the occupancy, I would set it as a property `atom.set("occupancy", site.occ)` or `atom.set("site_occupancy", site.occ)`.

Atomic number can not be overridden, so it could go in a property too `atom.set("cif_atomic_number", site.atomic_number())`

Reply:
Only occupancy is native to CIF format, the other two are actually determined based on element name by Gemmi… so I decided to only keep the occupancy.